[#69437] Fix Projects formatting in Webhooks table

myabc opened the pull request:
# Ticket

https://community.openproject.org/wp/69437

# What are you trying to accomplish?

Fixes display of Enabled Projects in Admin > Webhooks table.

After discussion with @HDinger, we agreed that we should:

- always show either a project count or "all projects" in the Enabled Projects column.
- always show the full list of resources in the Event Resources column.
- use default formatting rather than a badge (Primer Label/Counter) for counts, etc.

## Screenshots

### Before

<img width="993" height="462" alt="Screenshot 2025-12-17 at 12 32 33" src="https://github.com/user-attachments/assets/a1cea093-67a4-402a-9cf6-327ab75ddab7" />

### After

<img width="990" height="458" alt="Screenshot 2025-12-17 at 12 32 07" src="https://github.com/user-attachments/assets/41cabba7-9df9-4e58-a3b3-f4c262ba910c" />


# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
[#69437] Fix Projects formatting in Webhooks table

Changes behavior to:

- always show either a project count (without a badge/label) or "all
  projects" in the Enabled Projects column.
- always show the full list of resources in the Event Resources column.

myabc committed:
Add skeleton spec for Webhooks TableComponent

Adds shared examples for "generic tables", similar to Border Box Table
shared examples added in 54347731 but simplified.

myabc committed:
Allow localization of list delimiter

cbliard reviewed (commented):
Some non-code comments about the functionality.

Here is a screenshot of the ones I created locally:

<img width="1299" height="312" alt="image" src="https://github.com/user-attachments/assets/9ede0802-ab7e-47b9-8b5c-b021e6c60864" />

- I would prefer just "Events" instead of "Event resources" as only the name "Event" is used in the creation / edit pages, and in the description above
- The capitalization is different between "no projects" and "No event resources". It should be consistent. I think capital for both is better.
- When there is nothing (no projects / no events), text color could use something lighter like color muted, just so that it does not draw too much attention.
  - Or maybe it should draw attention because if nothing is selected, it's not doing anything and should be fixed. Maybe the whole row with a gray background? A little ⚠️ somewhere?
- I don't like truncated text where text can't be read. I see it as a design failure everytime. Why not propose a multiline?
- It lacks details about Update / Create being selected, why not add them?

Here is an example mockup:

<img width="1296" height="305" alt="image" src="https://github.com/user-attachments/assets/c79cbc9e-aeae-41cc-9ae0-bde04cac5d34" />

Also not related, but in the create/edit page, if the url is not valid, the page is displayed again but the selection of events is lost. It should be kept.

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/row_component_spec.rb:
```diff
@@ -0,0 +1,104 @@
+    render_inline(described_class.new(...))
+  end
+
+  let(:webhook) { build_stubbed(:webhook, name: "Hook me up", all_projects:) }
```
I think this `let(:webhook)` is too far away from the code.
It would be easier to read (at least for me) if it was closer to the test

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/row_component_spec.rb:
```diff
@@ -0,0 +1,104 @@
+  subject(:rendered_component) do
+    render_component(row: webhook, table:)
+  end
+
+  describe "Name" do
+    let(:column) { :name }
+
+    it "renders name" do
+      expect(rendered_component).to have_content "Hook me up"
+    end
+  end
```
For instance, here I find it reads better like this with the `webhook` being near the code, and removing the `all_projects` param as it adds noise for this particular test:

```suggestion
  describe "Name" do
    let(:webhook) { build_stubbed(:webhook, name: "Hook me up") }
    let(:column) { :name }

    it "renders name" do
      expect(rendered_component).to have_content "Hook me up"
    end
  end
```

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/row_component_spec.rb:
```diff
@@ -0,0 +1,104 @@
+  end
+
+  describe "Enabled Projects" do
+    let(:column) { :selected_projects }
+
+    context "when all projects enabled" do
+      it "renders 'All projects'" do
+        expect(rendered_component).to have_content "All projects"
+      end
+    end
```
making `webhook` closer to the code again.

```suggestion
    context "when all projects enabled" do
      let(:webhook) { build_stubbed(:webhook, all_projects: true) }
    
      it "renders 'All projects'" do
        expect(rendered_component).to have_content "All projects"
      end
    end
```

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/row_component_spec.rb:
```diff
@@ -0,0 +1,104 @@
+
+    context "when some projects are enabled" do
+      let(:projects) { create_list(:project, 2) }
+      let(:webhook) { create(:webhook, name: "Hook me up", all_projects: false, projects:) }
```
Remove noise (name not needed in this test, and `let(:projects)` not useful).
And I'd like to use `build_list` instead of `create_list` to continue using stubs, but it does not seem to work (and no time to investigate further).

```suggestion
      let(:webhook) { create(:webhook, all_projects: false, projects: create_list(:project, 2)) }
```

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/row_component_spec.rb:
```diff
@@ -0,0 +1,104 @@
+        expect(rendered_component).to have_content "2 projects"
+      end
+    end
+
+    context "when no projects are enabled" do
+      let(:all_projects) { false }
+
+      it "renders 'no projects'" do
+        expect(rendered_component).to have_content "no projects"
+      end
+    end
```
```suggestion
    context "when no projects are enabled" do
      let(:webhook) { create(:webhook, all_projects: false) }

      it "renders 'no projects'" do
        expect(rendered_component).to have_content "no projects"
      end
    end
```

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/row_component_spec.rb:
```diff
@@ -0,0 +1,104 @@
+
+  describe "Event Resources" do
+    let(:column) { :events }
+    let(:webhook) { create(:webhook, name: "Hook me up", event_names:) }
```
Same, I would remove that one `webhook` here, inline it in tests, remove the `name` attribute because that's noise, and inline the `let(:event_names)` to remove 1 level of indirection.

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/row_component_spec.rb:
```diff
@@ -0,0 +1,104 @@
+  let(:table) do
+    instance_double(Webhooks::Outgoing::Webhooks::TableComponent, columns: [column],
+                                                                  target_controller: "webhooks/outgoing/admin")
+  end
```
I think I would inline that `table` in `subject` too. It took me a long time to connect the dots between `let(:column) { :name }`, and seeing the `subject` definition not using that `column`. Then I saw it's used in `table`, which is a double, and then I finally saw that `table` is given as a parameter to `render_component`.

I think I find it easier to read when everything is at the same place. And split only if needed (add complexity if needed).

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/row_component_spec.rb:
```diff
@@ -0,0 +1,104 @@
+  let(:table) do
+    instance_double(Webhooks::Outgoing::Webhooks::TableComponent, columns: [column],
+                                                                  target_controller: "webhooks/outgoing/admin")
+  end
+  let(:all_projects) { true }
+
+  subject(:rendered_component) do
+    render_component(row: webhook, table:)
+  end
```
Something like this, with less indirections:

```suggestion
  subject(:rendered_component) do
    table = instance_double(Webhooks::Outgoing::Webhooks::TableComponent,
                            columns: [column],
                            target_controller: "webhooks/outgoing/admin")
    render_inline(described_class.new(row: webhook, table:))
  end

```

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/table_component_spec.rb:
```diff
@@ -0,0 +1,76 @@
+    include_examples "rendering generic table header", text: "Enabled projects"
+    include_examples "rendering generic table header", text: "Event resources"
+    include_examples "rendering generic table header", text: "Description"
+  end
```
I wonder if these tests have a lot of value.

A value of a test is in what its failure would teach us.

Here if it fails, it's either because:
- the rendering failed => learned a lot, we have a bug that needs fixing
- we changed the column names => it should not be a surprise, that's intended most of the time
- we added / removed a column => same, that's probably intended

That means that if we add or remove a column, the failing tests adds an additional maintenance burden rather than serving us. Except if adding / removing can be accidental in which case the test helps, but not sure.

The only usefulness I see is about doing a rendering and ensuring there is no failure doing so.

That's a bit theoretical. What do you think about it?

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/table_component_spec.rb:
```diff
@@ -0,0 +1,76 @@
+  context "with no webhooks" do
+    let(:webhook) { create_list(:webhook, 0) }
+
+    it_behaves_like "rendering generic table"
```
I think I would only keep that one, and I would parameterize it.

The test I'd like to write is this one:

 it_behaves_like "rendering generic table", expected_headers_count: 6, expected_rows_count:
```suggestion
    it_behaves_like "rendering generic table", expected_headers_count: 6,
                                              expected_rows_count: 1,
```

And on the other side:
```ruby
RSpec.shared_examples_for "rendering generic table" do |expected_headers_count: nil, expected_rows_count: nil|
  with = []
  with << "#{expected_headers_count} headers" if expected_headers_count
  with << "#{expected_rows_count} rows" if expected_rows_count
  with = with.any? ? " with #{with.to_sentence}" : ""

  it "renders generic table#{with}" do
    expect(rendered_component).to have_selector :table
    if expected_headers_count
      expect(rendered_component).to have_css("colgroup > col", count: expected_headers_count)
                               .and have_css("thead th", count: expected_headers_count)
    end
    if expected_rows_count
      expect(rendered_component).to have_css "tbody tr", count: expected_rows_count
    end
  end
end
```

output looks like this then:
```
Webhooks::Outgoing::Webhooks::TableComponent
  with no webhooks
    renders 'no webhooks' message
    behaves like rendering generic table
      renders generic table with 6 headers and 1 rows
  with webhooks
    behaves like rendering generic table
      renders generic table
```

cbliard reviewed (commented):
Some comments about the tests

myabc commented:
@cbliard thanks for the really useful feedback!

> The capitalization is different between "no projects" and "No event resources". It should be consistent. I think capital for both is better.

True. I also prefer the capitalised form. However in this case I wanted to reuse an existing translation (`label_x_projects`). This begs the question, would it be OK to call `t(:label_x_projects, count: webhook.projects.size).titleize`? Or might that cause issues in languages with different case rules like Turkish? (definitely an edge case - but something I raised when doing a recent review https://github.com/opf/openproject/pull/21398#discussion_r2606928872)

> When there is nothing (no projects / no events), text color could use something lighter like color muted, just so that it does not draw too much attention.

> Or maybe it should draw attention because if nothing is selected, it's not doing anything and should be fixed. Maybe the whole row with a gray background? A little ⚠️ somewhere?

I like the latter suggestion.

> I don't like truncated text where text can't be read. I see it as a design failure everytime. Why not propose a multiline?

I think multiline looks fine in this instance, although we might still want to clamp the number of lines (achievable with the widely-supported but still vendor prefixed `-webkit-line-clamp`). I just introduced a` Truncate` Stimulus controller in #21288 which provides the necessary behaviour for Primer's [`HiddenTextExpander`](https://primer.style/product/getting-started/rails/components/hidden_text_expander/). I may also see how that looks.

https://github.com/user-attachments/assets/fbed40ec-8754-429d-bb5c-b9d7d481e261

> It lacks details about Update / Create being selected, why not add them?

Good point.

myabc commented:
> I would prefer just "Events" instead of "Event resources" as only the name "Event" is used in the creation / edit pages, and in the description above

I'll also change the column header.

myabc committed:
Add rudimentary OpPrimer::WarningText component

`WarningText` renders `Primer::Text` with a leading alert Octicon using
the "attention" semantic color.

myabc committed:
UX/design improvements based on review feedback

- Removes truncation from Events column, instead showing Events grouped
  by Resource with each Resource on its own line.
- Shows warning styling when there are no Enabled Projects or Events.
  While validations currently allow this, it makes the webhook effectively
  pointless, so users should be alerted.
- Capitalizes "No projects" / "No event resources" consistently.

myabc committed:
Fix precedence in Octicon selector

myabc committed:
Add accessible filter set to Primer selectors

See 41417b91.

myabc committed:
Rename "Event Resources" column to "Events"

myabc committed:
Move explanation text to Page Header

myabc committed:
Autocorrect Rubocop offenses in EventResources

cbliard commented on modules/webhooks/config/locales/en.yml:
```diff
@@ -28,11 +28,10 @@ en:
-        other: "%{count} event resources"
-        zero: "No event resources"
+      label_x_events:
+        one: "1 events"
```
it should be "1 event"

cbliard replied:
I pushed a fix

cbliard commented on modules/webhooks/spec/components/webhooks/outgoing/webhooks/row_component_spec.rb:
```diff
@@ -0,0 +1,108 @@
+  def render_component(...)
+    render_inline(described_class.new(...))
+  end
+
```
These lines can now be deleted
```suggestion
```

cbliard reviewed (approved):
Nice 👏 

Good to merge for me.

cbliard committed:
Fix typo

myabc committed:
Remove unused method in RowComponent spec

Co-authored-by: Christophe Bliard <c.bliard@openproject.com>

cbliard commented:
I might be extra nitpicky, but the event names are sorted in the index view, while they are not in the edit / create views.
In English, it's ok, that's the same order in both screens, but in other languages, the order could be different.

See these screenshots with French:

index page

<img width="1203" height="457" alt="image" src="https://github.com/user-attachments/assets/a72fed94-2d97-437a-957b-1521509a7900" />

edit / create page

<img width="512" height="305" alt="image" src="https://github.com/user-attachments/assets/f2688725-5de6-4724-87e4-49c864c803aa" />

And I didn't know there was also a show page

<img width="643" height="232" alt="image" src="https://github.com/user-attachments/assets/9140ec2a-9a7a-41d6-be91-d947a2e1a2c5" />

The order is on show page is consistent with edit / create page

This would be better if that's consistent (but at the same time, that's not a big deal if it's not).
